semantic-ast/app/Main.hs

```diff
@@ -0,0 +1,145 @@
+    values, functions :: Traversal' Dec Name
+    values = _ValD . typed . _VarP
+    functions = _FunD . typed
+
+    truncate :: Name -> Name
+    truncate = mkName . nameBase
+
+parserForLanguage :: Language -> Ptr TreeSitter.Language.Language
+parserForLanguage = \case
+  Unknown -> error "Unknown language encountered"
+  CodeQL -> CodeQL.tree_sitter_ql
+  Go -> Go.tree_sitter_go
+  Haskell -> error "Haskell backend not implemented yet"
+  Java -> Java.tree_sitter_java
+  JavaScript -> TypeScript.tree_sitter_typescript
+  JSON -> JSON.tree_sitter_json
+  JSX -> TSX.tree_sitter_tsx
+  Markdown -> error "Markdown backend deprecated"
+  PHP -> PHP.tree_sitter_php
+  Python -> Python.tree_sitter_python
+  Ruby -> Ruby.tree_sitter_ruby
+  TypeScript -> TypeScript.tree_sitter_typescript
+  TSX -> TSX.tree_sitter_tsx
+
+main :: IO ()
+main = do
+  Config language path <- Opt.getRecord "generate-ast"
+  let lang = textToLanguage language
+  decls <- T.pack . pprint . fmap adjust <$> astDeclarationsIO (parserForLanguage lang) path
+  when (lang == Unknown) (die ("Couldn't determine language for " <> T.unpack language))
+
+  let programText =
+        [trimming|
+-- Language definition for $language, generated by ast-generate. Do not edit!
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE QuantifiedConstraints #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -fno-warn-unused-imports #-}
+
+module Language.$language.AST (module Language.$language.AST) where
+
+import qualified AST.Parse
+import qualified AST.Token
+import qualified AST.Traversable1.Class
+import qualified AST.Unmarshal
+import qualified Data.Foldable
+import qualified Data.List as Data.OldList
+import qualified Data.Maybe as GHC.Maybe
+import qualified Data.Text.Internal
+import qualified Data.Traversable
+import qualified GHC.Base
+import qualified GHC.Generics
+import qualified GHC.Records
+import qualified GHC.Show
+import qualified Prelude as GHC.Classes
+import qualified TreeSitter.Node
+
+debugSymbolNames :: [GHC.Base.String]
+debugSymbolNames = debugSymbolNames_0
+
+$decls
+  |]
+  hasOrmolu <- findExecutable "ormolu"
+  if isNothing hasOrmolu
+    then do
+      T.putStrLn programText
+      hPutStrLn stderr "generate-ast: No `ormolu` executable found, output will look janky."
+    else do
+      (path, tf) <- openTempFile "/tmp" "generated.hs"
+      T.hPutStrLn tf programText
+      hClose tf
+      callProcess "ormolu" ["--mode", "inplace", path]
```

Comment:
We could link in the `ormolu` package, at the cost of a ton of dependencies. It seems to me to be easier just to shell out in this invocation. The most idiomatic way to do this would be to use `streaming-process` and run ormolu such that it looks from `stdin`, but who cares? This is fine.